server/aws-lsp-codewhisperer/src/language-server/telemetryService.ts

```diff
@@ -0,0 +1,122 @@
+                break
+            case 'Reject':
+                suggestionState = 'REJECT'
+                break
+            case 'Discard':
+                suggestionState = 'DISCARD'
+                break
+            default:
+                suggestionState = 'EMPTY'
+        }
+        return suggestionState
+    }
+
+    private isLoginInvalidForTelemetry(): boolean {
+        return this.credentialsType === 'iam' || (this.loginType === 'builderId' && this.optOutPreference === 'OPTOUT')
+    }
+
+    private invokeSendTelemetryEvent(event: any) {
+        const request: SendTelemetryEventRequest = {
+            telemetryEvent: {
+                userTriggerDecisionEvent: event,
+            },
+        }
+        if (this.userContext !== undefined) {
+            request.userContext = this.userContext
+        }
+        if (this.optOutPreference !== undefined) {
+            request.optOutPreference = this.optOutPreference
+        }
+        this.sendTelemetryEvent(request)
+    }
+
+    public emitUserTriggerDecision(session: CodeWhispererSession, timeSinceLastUserModification?: number) {
+        if (this.isLoginInvalidForTelemetry()) {
+            return
+        }
+        const completionSessionResult = session.completionSessionResult ?? {}
+        const acceptedSuggestion = session.suggestions.find(s => s.itemId === session.acceptedSuggestionId)
+        const generatedLines =
+            acceptedSuggestion === undefined || acceptedSuggestion.content.trim() === ''
+                ? 0
+                : acceptedSuggestion.content.split('\n').length
+        const referenceCount =
+            acceptedSuggestion === undefined
+                ? 0
+                : acceptedSuggestion.references && acceptedSuggestion.references.length > 0
+                  ? 1
+                  : 0
+
+        const event: UserTriggerDecisionEvent = {
+            sessionId: session.codewhispererSessionId || '',
+            requestId: session.responseContext?.requestId || '',
+            customizationArn: session.customizationArn === '' ? undefined : session.customizationArn,
+            programmingLanguage: {
+                languageName: getRuntimeLanguage(session.language),
+            },
+            completionType:
+                session.suggestions.length > 0 ? getCompletionType(session.suggestions[0]).toUpperCase() : '',
+            suggestionState: this.getSuggestionState(session),
+            recommendationLatencyMilliseconds: session.firstCompletionDisplayLatency
+                ? session.firstCompletionDisplayLatency
+                : 0,
+            timestamp: new Date(Date.now()),
+            triggerToResponseLatencyMilliseconds: session.timeToFirstRecommendation,
+            suggestionReferenceCount: referenceCount,
+            generatedLine: generatedLines,
+            numberOfRecommendations: session.suggestions.length,
+            perceivedLatencyMilliseconds: timeSinceLastUserModification,
```

Comment:
Looking in VSCode `perceivedLatencyMilliseconds`, is computed based on triggerType https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/codewhisperer/util/codeWhispererSession.ts#L86-L92

We need to add case for manual trigger and use `timeToFirstRecommendation` here